source/fuzz/transformation_add_copy_memory.cpp

```diff
@@ -0,0 +1,180 @@
+  if (!fuzzerutil::IsFreshId(ir_context, message_.target_id())) {
+    return false;
+  }
+
+  // Check that instruction descriptor is valid.
+  auto* inst = FindInstruction(message_.instruction_descriptor(), ir_context);
+  if (!inst) {
+    return false;
+  }
+
+  // Check that we can insert OpCopyMemory before |instruction_descriptor|.
+  auto iter = fuzzerutil::GetIteratorForInstruction(
+      ir_context->get_instr_block(inst), inst);
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(SpvOpCopyMemory, iter)) {
+    return false;
+  }
+
+  // Check that source instruction exists.
+  auto* source_inst =
+      ir_context->get_def_use_mgr()->GetDef(message_.source_id());
+  if (!source_inst) {
+    return false;
+  }
+
+  // Check that result type of source instruction exists and can be used with
+  // OpCopyMemory.
+  const auto* source_type =
+      ir_context->get_type_mgr()->GetType(source_inst->type_id());
+  if (!source_type || !source_type->AsPointer()) {
+    return false;
+  }
+
+  if (!CanUsePointeeWithCopyMemory(*source_type->AsPointer()->pointee_type())) {
+    return false;
+  }
+
+  // OpTypePointer with Private storage class exists.
+  if (!fuzzerutil::MaybeGetPointerType(
+          ir_context,
+          ir_context->get_type_mgr()->GetId(
+              source_type->AsPointer()->pointee_type()),
+          SpvStorageClassPrivate)) {
+    return false;
+  }
+
+  // Check that this transformation respects domination rules.
+  const auto* source_block = ir_context->get_instr_block(source_inst);
+  const auto* target_block = ir_context->get_instr_block(inst);
+  assert(source_block && target_block);
+
+  // Check that both source and target instructions are in the same function.
+  if (source_block->GetParent() != target_block->GetParent()) {
+    return false;
+  }
+
+  // Check domination rules.
+  return source_inst != inst &&
+         ir_context->GetDominatorAnalysis(source_block->GetParent())
+             ->Dominates(source_inst, inst);
+}
+
+void TransformationAddCopyMemory::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  // Result id for target variable type.
+  // TODO: it would be good to refactor variable creation part into a separate
+  //  function (in, say, fuzzerutil). This will reduce boilerplate here, in
+  //  TransformationPushIdsThroughVariables and two transformations that create
+  //  variables.
+  auto type_id = fuzzerutil::MaybeGetPointerType(
+      ir_context,
+      fuzzerutil::GetPointeeTypeIdFromPointerType(
+          ir_context, fuzzerutil::GetTypeId(ir_context, message_.source_id())),
+      SpvStorageClassPrivate);
+
+  // Create global target variable.
+  opt::Instruction::OperandList variable_operands = {
+      {SPV_OPERAND_TYPE_STORAGE_CLASS, {SpvStorageClassPrivate}}};
+  ir_context->AddGlobalValue(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpVariable, type_id, message_.target_id(),
+      std::move(variable_operands)));
```

Comment:
Does the target have to be global?

How about you add a StorageClass field to the transformation that can either be Function or Private?

Then, depending on that storage class, you add a variable to the function or a global variable (taking account of entry point interfaces in the latter case)?

Reply:
Fixed in 821685d.